GTC 1862 spike pagination

gtempus opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request, or let us know here if you need any help -->

## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Make sure you are requesting to pull a topic/feature/bugfix branch (right side). Don't request your master!
- [x] Make sure you are making a pull request against the develop branch (left side). Also you should start your branch off our develop.
- [x] Check the commit's or even all commits' message styles matches our requested structure.
- [ ] Check your code additions will fail neither code linting checks nor unit test.



## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. --> 

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. --> 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [x] Other (please describe): This is a spike to learn how to use the [fastapi-pagination](https://github.com/uriyyo/fastapi-pagination) package


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
No pagination on the `datasets` endpoint
Issue Number: N/A


## What is the new behavior?
<!-- Please describe the behavior or changes that are being added by this PR. -->
Pagination is now happening on the test endpoint: `/datasets?page=1&size=10` 
-
-
-

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->
This PR is not intended for merging. It's just an experiment to learn about fastapi-paginate.
However, I still wanted to avoid breaking the existing functionality. Therefore, I did add an extra route and small function to the `crud/datasets` module.

The response model is not sufficient for production. It was just a way to get something coming back from the server.

Manual Testing:
- Original behavior endpoint: `/datasets`
  - The result is unchanged. No new pagination fields are included.
- To enable pagination: `/datasets?page=1&size=1`
  - `page` needs to be greater than `0`
  - `size` can be between `1` and `49`, inclusive*
  - Now there are `links` and `meta` properties included in the response.
*I have some silly logic right now that decides to paginate whether `size < 50`

Initial thoughts:
- SQLAlchemy - The install of fastapi-pagination did cause the update of several packages. I had to introduce tighter constraints on `sqlalchemy` and pin it to `1.3.x` because it seemed that `1.4.x` was breaking the API.
- We'll need a way to get to the SQLAlchemy Query object before causing gino to execute the query. It's not a dealbreaker but some refactoring of the crud modules will need to happen for the proposed endpoints.
- Some consideration will need to go into extending the Page and/or DatasetsResponse types to bring them into alignment with Resource Watch. Extension/Customization steps are outlined in the [fastapi-pagination repo](https://github.com/uriyyo/fastapi-pagination)

gtempus committed:
fix: Add missing migrations from feature/Revisions branch

GTC-1862

gtempus committed:
spike: Easiest possible integration with fastapi-paginate

This branch is not intended for merging into main. It's just an experiment to learn about fastapi-paginate.
However, I still wanted to avoid breaking the existing functionality. Therefore, I did add an extra route and small function to the `crud/datasets` module.

The response model is not sufficient for production. It was just a way to get something coming back from the server.

Manual Testing:
- You can hit this endpoint: `/datasets/paginate?page=4&size=1`
- `page` needs to be greater than `0`
- `size` can be between `1` and `100`, inclusive

Initial thoughts:
- SQLAlchemy - The install of fastapi-pagination did cause the update of several packages. I had to introduce tighter constraints on `sqlalchemy` and pin it to `1.3.x` because it seemed that `1.4.x` was breaking the API.
- We'll need a way to get to the SQLAlchemy Query object before causing gino to execute the query. It's not a dealbreaker but some refactoring of the crud modules will need to happen for the proposed endpoints.
- Some consideration will need to go into extending the Page and/or DatasetsResponse types to bring them into alignment with Resource Watch. Extension/Customization steps are outlined in the (fastapi-pagination repo)[https://github.com/uriyyo/fastapi-pagination]

GTC-1862

gtempus committed:
fix: Pin some dependencies because of breaking tests.

GTC-1862

jterry64 commented:
@gtempus I don't foresee any breaking changes with the package updates, as long as tests are passing.

This is cool! Do you know why the response has the __value__ and __profile__ keys for each object, and is that something we can customize? Seems like it might it be like Python hidden attribute names based on the underscores?

And yeah, if we're going to use the page number/size method, we should probably align with RW. 

I haven't implemented pagination on anything in awhile, but based on the amount of customization we might need to do for this library, would it be worth it vs just implementing pagination ourselves? I.e., does this library provide good value?

Those are just my initial thoughts.s

gtempus committed:
feat: Sketch of the suggested pagination approach

This is the closest I could come to my original idea of 'decorating' our existing code to inject pagination. I'm particularly pleased that I didn't have to change pydantic response types (this due to the custom `@partial` python decorator) on the newly defined `PaginatedResponse`.

GTC-1862

gtempus committed:
feat: Refine response types

I did notice that `success` isn't being populated by default because of the `@partial` decorator's use of `exclude_unset`. It's not great that we have to explicitly set this value in the route, but I think it's still worth going this direction.

GTC-1862

gtempus commented on app/routes/datasets/datasets.py:
```diff
@@ -16,9 +17,17 @@
     response_model=DatasetsResponse,
 )
-async def get_datasets() -> DatasetsResponse:
+async def get_datasets(params: Params = Depends()) -> DatasetsResponse:
     """Get list of all datasets."""
+    if params.size < 50:
+        query = datasets.get_datasets_query()
+        data = await paginate(query, params)
+        adapted = [dataset.__values__ for dataset in data.items]
+        return DatasetsResponse(data=adapted,
+                                    links={'self': "", 'first': "", 'last': "", 'prev': "", 'next': ""},
+                                    meta={'size': data.size, 'total-pages': "", 'total-items': data.total},
+                                    status='success')
 
     data = await datasets.get_datasets()
+    return DatasetsResponse(data=data, status='success')
```
This actually turned out to be the hardest part for introducing a breaking changes _without_ actually breaking things. The main idea was to have the same endpoint _do the right thing_ if pagination params were given but otherwise remain unchanged to preserve legacy behavior.

With `pydantic` I couldn't just create subclasses of `DatasetsResponse` and continue to return the generalized one. When I did actually tried to do that, `pydantic` would throw a ValidationError and tell me that I was either including extra fields or it would add the pagination properties when they should've been hidden.

gtempus commented on app/routes/datasets/datasets.py:
```diff
@@ -16,9 +17,17 @@
+        data = await paginate(query, params)
+        adapted = [dataset.__values__ for dataset in data.items]
+        return DatasetsResponse(data=adapted,
+                                    links={'self': "", 'first': "", 'last': "", 'prev': "", 'next': ""},
+                                    meta={'size': data.size, 'total-pages': "", 'total-items': data.total},
```
Pydantic types for these would be created in the actual implementation.

gtempus commented on app/models/pydantic/responses.py:
```diff
@@ -1,8 +1,75 @@
+    return decorator
+
+
+@partial('links', 'meta')
```
I did some research to see how other folks were solving similar issues of how to get `pydantic` to completely remove various properties. My implementation is a working refinement of the [thoughts in this pydantic thread](https://github.com/samuelcolvin/pydantic/issues/1223). 

It's a decorator that lets me define properties that should not exists (in any way) if they are not explicitly passed in during initialization. That way, there are no new properties revealed when legacy list endpoints are called (so we don't inadvertently break existing clients). 

It's not perfect. Because of the `exclude_unset` directive, we would need to pass the `success` property explicitly now from the route code. I don't think that's a deal breaker.

gtempus committed:
refactor: Use Union instead of `@partial` decorator

@dmannarino suggested the `Union` type. It's much simpler. I had 'tunnel vision' regarding the function return type and the pydantic `response_model` parameter.

GTC-1862

gtempus reviewed (commented):
After walking though the code with @dmannarino we made these decisions:
### Response Types
- Use `Union` to explicitly show that one of two types of responses could happen (paginated or _not_ paginated for legacy support)
- Make `pydantic` types for the `links` and `meta` properties within the paginated response.
### Pagination Implementation
- Use [fastapi-pagination](https://github.com/uriyyo/fastapi-pagination)
- Design our usage around it so that we can easily replace it, if necessary.
- Spend some effort separating query building from execution so that pagination can be injected cleanly into the existing crud modules
### Documentation
- Ensure that any pagination information is clear in the API docs
